User Management for GitLab via Artemis

ungaralex opened the pull request:
<!-- Thanks for contributing to Artemis! Before you submit your pull request, please make sure to check the following boxes by putting an x in the [ ] (don't: [x ], [ x], do: [x]) -->
<!-- If your pull request is not ready for review yet, create a draft pull request! -->

### Checklist
- [x] I tested *all* changes and *all* related features with different users (student, tutor, instructor, admin) on the test server https://artemistest.ase.in.tum.de.
- [x] Server: I implemented the changes with a good performance and prevented too many database calls
- [x] Server: I documented the Java code using JavaDoc style.
- [x] Client: I documented the TypeScript code using JSDoc style.
- [x] Client: I added multiple screenshots/screencasts of my UI changes
- [x] Client: I translated all the newly inserted strings into German and English

### Motivation and Context
<!-- Why is this change required? What problem does it solve? -->
<!-- If it fixes an open issue, please link to the issue here. -->
**This PR depends on #1110**
For our GitLab+Jenkins setup we don't have any external user management setup like Jira atm. Because of that, we have to manage all users and groups using the Artemis DB.

### Description
<!-- Describe your changes in detail -->
- I extended the user management UI allowing admins to also specify the password and edit the groups a user belongs to Those groups now also get displayed in the user component next to the roles. 
- The server now also provides an `ArtemisInternalAuthenticationProvider` in case `jira` and `ldap` are not activated as a Spring profile. In this case, all users get authenticated and managed using the internal Artemis DB. 
- The GitLab service and VCS interface now also allow Artemis to manage and sync the users and groups. This means that any new user gets also added to GitLab and any updated user (e.g. new password) gets also updated in GitLab. Since we also allow instructors to change the groups of a course after creating it, if there are already exercises in this course, the related repositories and groups in GitLab will also get updated (adding/removing new/old instructors/tutors, upgrading/downgrading instructors/tutors). The same goes for changed groups of a user itself. If a user should not be part of an instructor/tutor group after the update, the related entities in GitLab will also be changed.

### Steps for Testing
<!-- Please describe in detail how the reviewer can test your changes. -->

0. Message me in order to get an account for the testserver
1. Log in to the Artemis Testserver 2
2. Create some new users
3. Those users should also get created in GitLab
4. Create a new course and a programming exercise
5. The users should have the correct rights for the new programming exercise in GitLab (instructors --> Maintainer, tutors --> Guest)
6. Add some more users (tutors, instructors)
7. Those users should now also be a member of the programming exercise group in GitLab
8. Add some users that do not belong to any instructor/tutor group
9. Change the groups of the course, so that some of the new users should now be an instructor/tutor
10. The groups in GitLab should be updated
11. Change the groups of the course, so that some instructors should now only be tutors and check GitLab
12. Change the groups of the course, so that some tutors should now be instructors and check GitLab
13. Change the groups of the course, so that some users are now longer an instructor or tutor and check GitLab

### Screencast
![gitla_user_management](https://user-images.githubusercontent.com/27979674/72122292-b29f9780-335d-11ea-8e52-9f1657e96de8.gif)

krusche commented:
Thanks for getting this done. I think there are some changes that can break a lot of behavior, so I think it would be great if you can add some test cases, at least the minimal ones for the functionality that you have changed in this PR so that we really know that nothing important broke.

ungaralex commented:
I only changed the behavior of the update/create methods in the `UserService` and added some additional logic in case we have no external user management like Jira.

Based on that I added some tests:
- Test the creation of new users
- Test the update of existing users
- Verify, that the (user management portion of the) `VersionControlService` is **not** called for our default configuration (i.e. Bitbucket with Jira). This would be the case for the user create/update/delete operations as well as for the course update endpoint.
- Verify that the new `ArtemisInternalAuthenticationProvider` bean is not available for our default configuration, so it cannot be injected and cause unwanted behavior.

ungaralex commented:
@krusche If this PR is too big, I can split this up between the client and server changes.

krusche commented on package.json:
```diff
@@ -56,7 +56,7 @@
         "ng2-dnd": "5.0.2",
         "ngx-cacheable": "1.3.2",
-        "ngx-chips": "2.1.0",
+        "ngx-chips": "^2.1.0",
```
please revert this change!

ungaralex replied:
Reverted

krusche commented on src/main/java/de/tum/in/www1/artemis/web/rest/CourseResource.java:
```diff
@@ -195,7 +202,12 @@ public CourseResource(Environment env, UserService userService, CourseService co
                     return ResponseEntity.badRequest().headers(HeaderUtil.createAlert(applicationName, "The shortname is invalid", "shortnameInvalid")).body(null);
                 }
                 checkIfGroupsExists(updatedCourse);
+                final var oldInstructorGroup = existingCourse.get().getInstructorGroupName();
```
please add some comments what this code wants to achieve. It's not really clear to me at the moment why we need this here

ungaralex replied:
I added some comments.
Basically, we need to calculate the diff between the old instructors/tutors and the new ones. Based on that, the `VCSUserManagementService` can then remove/add/update users for all exercise repositories in the related course.

krusche commented on src/main/java/de/tum/in/www1/artemis/web/rest/UserResource.java:
```diff
@@ -66,16 +69,26 @@
 
     private final UserService userService;
 
+    private final Optional<ContinuousIntegrationService> continuousIntegrationService;
```
I am not sure if it is a good idea to have these dependencies in the Resource. I guess we should move all this logic into the UserService

ungaralex replied:
Yes, this was a bad design choice. I moved the relevant code to the service

krusche commented on src/main/java/de/tum/in/www1/artemis/security/JiraAuthenticationProvider.java:
```diff
@@ -97,24 +102,27 @@ public Authentication authenticate(Authentication authentication) throws Authent
+            final var path = JIRA_URL + "/rest/api/2/user?username=" + username + "&expand=groups";
+            final HttpEntity<Principal> entity;
+            if (skipPasswordCheck) {
+                authenticationResponse = restTemplate.exchange(path, HttpMethod.GET, null, Map.class);
```
this code does not have the same logic as before.
`entity  = new HttpEntity<>(HeaderUtil.createAuthorization(JIRA_USER, JIRA_PASSWORD)):` is missing. I guess this is the case, because the REST template already includes this authorization. I wonder however what happens with the REST template if the `else` branch is invoked. Does that change the default authorization?

Also, if the code is correct, you could move `entity` into the else branch completely.

krusche replied:
I guess some comments would be helpful here as well!

ungaralex replied:
Yes, the authorization gets injected into the `RestTemplate` (Artemis Admin auth). In the else case, we use the user's credentials, because we don't skip the password check. I added some comments to clarify this.

krusche commented on src/main/java/de/tum/in/www1/artemis/security/JiraAuthenticationProvider.java:
```diff
@@ -215,11 +223,9 @@ public boolean supports(Class<?> authentication) {
         body.put("name", username);
-        HttpEntity<?> entity = new HttpEntity<>(body, headers);
-        RestTemplate restTemplate = new RestTemplate();
+        HttpEntity<?> entity = new HttpEntity<>(body);
```
For the new usage of this in JIRA, I think we should call `isGroupAvailable` before to throw a better error.
The super admin can enter any group name in the user management, and the group might not exist in JIRA.

Another option (instead of throwing an error) would be to create the group in such a case.

krusche commented on src/main/java/de/tum/in/www1/artemis/service/UserService.java:
```diff
@@ -347,33 +369,52 @@ public void updateUser(String firstName, String lastName, String email, String l
+        return user;
+    }
+
+    private void forwardUserUpdates(User user, Set<String> oldGroups) {
```
this name sounds strange. What does it mean? please find a more descriptive one
I guess JavaDoc would be good here.

ungaralex replied:
I changed the name and added some JavaDoc

krusche commented on src/main/java/de/tum/in/www1/artemis/service/UserService.java:
```diff
@@ -347,33 +369,52 @@ public void updateUser(String firstName, String lastName, String email, String l
      * @param login user login string
      */
     public void deleteUser(String login) {
+        optionalVcsUserManagementService.ifPresent(userManagementService -> userManagementService.deleteUser(login));
```
This should not delete a user in the external user management system (e.g. JIRA).

krusche replied:
I guess some comments would be good here as well

ungaralex replied:
This only deletes the user in the VCS and only if the bean is present (which is the case for GitLab, but not for Bitbucket). I also added some comments

krusche commented on src/test/resources/config/application.yml:
```diff
@@ -2,6 +2,7 @@ artemis:
     repo-clone-path: ./repos/
     repo-download-clone-path: ./repos-download/
     encryption-password: <encrypt-password>     # arbitrary password for encrypting database values
+    external-user-management: true
```
this is still outdated, please update the file accordingly

ungaralex replied:
Updated

krusche commented on src/main/java/de/tum/in/www1/artemis/service/connectors/gitlab/GitLabUserManagementService.java:
```diff
@@ -0,0 +1,271 @@
+
+    @Override
+    public void updateCoursePermissions(Course updatedCourse, String oldInstructorGroup, String oldTeachingAssistantGroup) {
+        if (oldInstructorGroup.equals(updatedCourse.getInstructorGroupName()) && oldTeachingAssistantGroup.equals(updatedCourse.getTeachingAssistantGroupName())) {
```
this method and the one below look quite complex to me. Please add more comments so that it is easier to understand what is happening and more importantly why this is happening!

ungaralex replied:
I added lots of new comments. Let me know, what you think

krusche commented on src/main/java/de/tum/in/www1/artemis/security/JiraAuthenticationProvider.java:
```diff
@@ -134,14 +145,14 @@ else if (e.getStatusCode().is5xxServerError()) {
+            final var content = authenticationResponse.getBody();
+            final var login = content.getName();
+            final var emailAddress = content.getEmailAddress();
+            final var user = userRepository.findOneByLogin(login).orElseGet(() -> userService.createUser(login, password, content.getDisplayName(), "", emailAddress, null, "en"));
```
```suggestion
            final var user = userRepository.findOneByLogin(login).orElseGet(() -> userService.createUser(login, "", content.getDisplayName(), "", emailAddress, null, "en"));
```

krusche replied:
we don't want to store the password of JIRA users in our db!!!

ungaralex replied:
reverted
